### PR TITLE
evalengine: Optimize NullSafeComparison

### PR DESCRIPTION
**Pull request by vmg**
## Description

Following up on @mattlord's https://github.com/vitessio/vitess/pull/9679, where we're starting to use our native Collation comparison code in more places, this PR optimizes `evalengine.NullSafeComparison` for the most common use cases (strings and numerics). 

To optimize string comparisons I worked on the `Collate` method in our native collations implementation, particularly the one for the `utf8mb4_0900_ai_ci`, `utf8mb4_0900_ai_cs` and `utf8mb4_ai_ci` collations, which are the default ones in MySQL 8.0+:

```
name                                              old time/op  new time/op  delta
UCA900Collation/Long/1/utf8mb4_0900_as_cs-16      6.16µs ± 1%  2.48µs ± 2%  -59.82%  (p=0.008 n=5+5)
UCA900Collation/Long/1/utf8mb4_0900_as_ci-16      6.12µs ± 1%  2.43µs ± 3%  -60.26%  (p=0.008 n=5+5)
UCA900Collation/Long/1/utf8mb4_0900_ai_ci-16      3.07µs ± 1%  1.32µs ± 1%  -56.96%  (p=0.008 n=5+5)
UCA900Collation/Long/8/utf8mb4_0900_as_cs-16      47.8µs ± 1%  18.3µs ± 2%  -61.84%  (p=0.008 n=5+5)
UCA900Collation/Long/8/utf8mb4_0900_as_ci-16      47.5µs ± 2%  18.3µs ± 1%  -61.36%  (p=0.008 n=5+5)
UCA900Collation/Long/8/utf8mb4_0900_ai_ci-16      23.7µs ± 1%   9.4µs ± 2%  -60.33%  (p=0.008 n=5+5)
UCA900Collation/Spanish/1/utf8mb4_0900_as_cs-16   5.54µs ± 1%  1.01µs ± 3%  -81.84%  (p=0.008 n=5+5)
UCA900Collation/Spanish/1/utf8mb4_0900_as_ci-16   5.44µs ± 1%  0.92µs ± 1%  -83.11%  (p=0.008 n=5+5)
UCA900Collation/Spanish/1/utf8mb4_0900_ai_ci-16   2.74µs ± 2%  0.52µs ± 4%  -80.85%  (p=0.008 n=5+5)
UCA900Collation/Spanish/8/utf8mb4_0900_as_cs-16   42.8µs ± 1%   4.3µs ± 2%  -89.92%  (p=0.008 n=5+5)
UCA900Collation/Spanish/8/utf8mb4_0900_as_ci-16   42.4µs ± 1%   4.2µs ± 1%  -89.98%  (p=0.008 n=5+5)
UCA900Collation/Spanish/8/utf8mb4_0900_ai_ci-16   21.1µs ± 2%   2.5µs ± 1%  -88.01%  (p=0.008 n=5+5)
UCA900Collation/English/1/utf8mb4_0900_as_cs-16   9.89µs ± 2%  0.45µs ± 1%  -95.45%  (p=0.008 n=5+5)
UCA900Collation/English/1/utf8mb4_0900_as_ci-16   9.81µs ± 1%  0.44µs ± 4%  -95.52%  (p=0.008 n=5+5)
UCA900Collation/English/1/utf8mb4_0900_ai_ci-16   4.93µs ± 2%  0.25µs ± 4%  -94.84%  (p=0.008 n=5+5)
UCA900Collation/English/8/utf8mb4_0900_as_cs-16   78.1µs ± 1%   2.2µs ± 1%  -97.17%  (p=0.008 n=5+5)
UCA900Collation/English/8/utf8mb4_0900_as_ci-16   77.9µs ± 1%   2.2µs ± 1%  -97.15%  (p=0.008 n=5+5)
UCA900Collation/English/8/utf8mb4_0900_ai_ci-16   38.7µs ± 1%   1.1µs ± 1%  -97.04%  (p=0.008 n=5+5)
UCA900Collation/Japanese/1/utf8mb4_0900_as_cs-16  13.5µs ± 1%  12.5µs ± 1%   -7.98%  (p=0.008 n=5+5)
UCA900Collation/Japanese/1/utf8mb4_0900_as_ci-16  9.06µs ± 2%  8.38µs ± 3%   -7.53%  (p=0.008 n=5+5)
UCA900Collation/Japanese/1/utf8mb4_0900_ai_ci-16  4.61µs ± 1%  4.26µs ± 1%   -7.52%  (p=0.008 n=5+5)
UCA900Collation/Japanese/8/utf8mb4_0900_as_cs-16   106µs ± 1%    99µs ± 1%   -6.93%  (p=0.008 n=5+5)
UCA900Collation/Japanese/8/utf8mb4_0900_as_ci-16  71.0µs ± 1%  64.9µs ± 3%   -8.63%  (p=0.008 n=5+5)
UCA900Collation/Japanese/8/utf8mb4_0900_ai_ci-16  36.1µs ± 1%  33.5µs ± 2%   -7.14%  (p=0.008 n=5+5)
```

The optimization uses fast weight tables, like our previous WeightString optimization, to speed up the comparison of strings that are composed of mostly ASCII codepoints, like English text usually this. The runtime of these comparisons has been reduced by up to 98%. The `Long` and `Spanish` benchmark cases verify that that optimization still applies for strings that have non-ASCII characters in sporadic places, such as Spanish accents or Emoji in English text. The `Japanese` benchmark case verifies the heuristic in our optimization, wherein text that is overwhelmingly non-ASCII falls back to the old comparison algorithm without slowing down with superfluous ASCII checks (there is, in fact, a small performance gain because I've inlined some calls to the old algorithm).

Once the collations comparison was optimized, I worked on the actual `NullSafeComparison` call, which had a lot of low-hanging fruit. The most notable optimization was removing all allocations for textual comparisons (a very frequent use case that shouldn't need to allocate), and use a memory pool to amortize allocations for numerical comparisons (a less frequent case that does need to allocate).

These are the results:

```
name                                                 old time/op    new time/op    delta
NullSafeComparison/Strings/utf8mb4_0900_ai_ci/1-16     23.7µs ± 1%     7.8µs ± 1%   -66.99%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_ai_ci/16-16    62.8µs ± 3%    21.4µs ± 3%   -65.87%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_as_cs/1-16     29.9µs ± 2%    10.3µs ± 1%   -65.48%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_as_cs/16-16     120µs ± 7%      42µs ± 1%   -65.15%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_general_ci/1-16     12.9µs ± 1%     2.6µs ± 1%   -79.59%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_general_ci/16-16    25.6µs ± 1%     9.2µs ± 3%   -63.98%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_bin/1-16       10.6µs ± 1%     1.3µs ± 2%   -87.37%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_bin/16-16      10.6µs ± 1%     1.3µs ± 3%   -87.37%  (p=0.008 n=5+5)
NullSafeComparison/Numeric-16                          62.2µs ± 1%    27.7µs ± 2%   -55.49%  (p=0.008 n=5+5)

name                                                 old alloc/op   new alloc/op   delta
NullSafeComparison/Strings/utf8mb4_0900_ai_ci/1-16     5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_ai_ci/16-16    5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_as_cs/1-16     5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_as_cs/16-16    5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_general_ci/1-16     5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_general_ci/16-16    5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_bin/1-16       5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_bin/16-16      5.76kB ± 0%    0.00kB       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Numeric-16                          23.1kB ± 0%     0.0kB           ~     (p=0.079 n=4+5)

name                                                 old allocs/op  new allocs/op  delta
NullSafeComparison/Strings/utf8mb4_0900_ai_ci/1-16       72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_ai_ci/16-16      72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_as_cs/1-16       72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_as_cs/16-16      72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_general_ci/1-16       72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_general_ci/16-16      72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_bin/1-16         72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Strings/utf8mb4_0900_bin/16-16        72.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
NullSafeComparison/Numeric-16                             288 ± 0%         0       -100.00%  (p=0.008 n=5+5)
```

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/collations/uca_test.go [hunk 1]
```diff
@@ -239,6 +240,10 @@ const ChineseString2 = "春江潮水连海平，海上明月共潮生。" +
 	"不知江月待何人，但见长江送流水。" +
 	"白云一片去悠悠，青枫浦上不胜愁。" +
 	"谁家今夜扁舟子？何处相思明月楼？"
+const SpanishString = "A mí se me hace cuento que empezó Buenos Aires: " +
```
Comment (systay) [inline go/mysql/collations/uca_test.go:243]: :argentina: !

File: go/vt/vtgate/evalengine/comparisons.go [hunk 2]
```diff
@@ -382,3 +386,165 @@ func (l *LikeExpr) typeof(env *ExpressionEnv) sqltypes.Type {
...
+		case result < 0:
+			return -1, nil
+		case result > 0:
+			return 1, nil
+		default:
+			return 0, nil
+		}
+
+	case sqltypes.IsNumber(typ):
+		v1cast := borrow()
```
Comment (systay) [inline go/vt/vtgate/evalengine/comparisons.go:472]: nit: `borrow`/`Put()`?

**Review by systay — APPROVED** [review]
This is really great! We're enabling more and more functionality that needs to run in the vtgate, so these speedups are critical. :heart_eyes_cat:

File: go/vt/vtgate/evalengine/comparisons.go (hunk 2 above)
Comment (vmg) [reply go/vt/vtgate/evalengine/comparisons.go]: yup cleaned up

**Pull request merged**